kapitan/refs/secrets/gsm.py

```diff
@@ -0,0 +1,145 @@
+"GCP secret manager secret retrieval module"
+import os
+import base64
+import re
+import json
+import hashlib
+from binascii import Error as b_error
+
+from kapitan.errors import KapitanError
+from kapitan import cached
+from kapitan.refs.base import RefError
+from kapitan.refs.base64 import Base64Ref, Base64RefBackend
+from kapitan.refs.base import REF_TOKEN_SUBVAR_PATTERN
+
+try:
+    from google.cloud import secretmanager
+except:
+    pass
+
+
+class GoogleSMError(KapitanError):
+    " Generic Google Secret Manager error "
+    pass
+
+
+def gsm_obj():
+    if not cached.gsm_obj:
+        gsm_client = secretmanager.SecretManagerServiceClient()
+        cached.gsm_obj = gsm_client
+    return cached.gsm_obj
+
+
+class GoogleSMSecret(Base64Ref):
+    """
+    Google Secret Manager (read only) support
+    """
+
+    def __init__(self, data, project_id, **kwargs):
+
+        self.data = data
+        self.project_id = project_id
+        self.version_id = "latest"
+        super().__init__(self.data, **kwargs)
+        self.type_name = "gsm"
+
+    @classmethod
+    def from_params(cls, data, ref_params):
+        """
+        GSM Secret type is read only and cannot generate a secret
```

Comment:
I think to add this functionality it's going to need a new operator for the secondary functions. Currently `gsm` just stores the encoded `secret_id` in the ref path wich it later uses to retrieve the secret from secret manager. The secondary functions just return one value i.e the secret. But to upload the secret to secret manager we would also need the `secret_id`. Maybe we can have a  secondary function separated by `&&` which returns a dictionary with both `secret_id` and secret. e.g `?{gsm:secret/path:version_id}||ramdomstr&&reveal:path/to/secret}`. here secret_id would be a `randomstr` and the secret would be referred from `reveal:path/to/secret`. Or am I missing an easier alternative?

Reply:
In case we want to keep it as read-only, we can create a new secondary function, say `secretid`, and add the secret_id value to it separated by `:`. e.g `?{gsm:secret/path:version_id||secretid:topsecret}`. This might make it easier for users to add secret ids (?)